torch/csrc/distributed/rpc/tensorpipe_agent.h

```diff
@@ -18,9 +18,61 @@ namespace rpc {
 using steady_clock_time_point =
     std::chrono::time_point<std::chrono::steady_clock>;
 
+struct TransportRegistration {
+  std::shared_ptr<tensorpipe::transport::Context> transport;
+  int64_t priority;
+  std::string address;
+};
+
+C10_DECLARE_REGISTRY(TensorPipeTransportRegistry, TransportRegistration);
+
+struct ChannelRegistration {
+  std::shared_ptr<tensorpipe::channel::Context> channel;
+  int64_t priority;
+};
+
+C10_DECLARE_REGISTRY(TensorPipeChannelRegistry, ChannelRegistration);
+
+constexpr auto kDefaultNumWorkerThreads = 16;
+
 struct TensorPipeRpcBackendOptions : public RpcBackendOptions {
-  TensorPipeRpcBackendOptions(float rpc_timeout, std::string init_method)
-      : RpcBackendOptions(rpc_timeout, init_method) {}
+  TensorPipeRpcBackendOptions(
+      int numWorkerThreads,
+      optional<std::vector<std::string>> transports,
+      optional<std::vector<std::string>> channels,
+      float rpc_timeout,
+      std::string init_method)
+      : RpcBackendOptions(rpc_timeout, init_method),
+        numWorkerThreads(numWorkerThreads),
+        transports(std::move(transports)),
+        channels(std::move(channels)) {
+    TORCH_CHECK(
+        numWorkerThreads > 0,
+        "num_worker_threads must be positive, got ",
+        numWorkerThreads);
+
+    if (transports.has_value()) {
+      for (const std::string& transportName : transports.value()) {
+        TORCH_CHECK(
+            TensorPipeTransportRegistry()->Has(transportName),
+            "Unknown transport: ",
+            transportName);
+      }
+    }
+
+    if (channels.has_value()) {
+      for (const std::string& channelName : channels.value()) {
+        TORCH_CHECK(
+            TensorPipeChannelRegistry()->Has(channelName),
+            "Unknown channel: ",
+            channelName);
+      }
+    }
```

Comment:
Do we need to verify transports and channels match?

Reply:
There's pros and cons. The pros of verifying is that we would catch a mistyped name and avoid it silently slipping through. The cons are that if we ever remove a backend then all the users configs that specify it will become invalid. These are private options and are intended to be short-lived (until we find a better long-term solution), so personally I feel the former concern outweighs the latter. What do you think?

Reply:
Sure, let's then keep this as-is.